File: python/leetcode/238.py
class Solution:
    def productExceptSelf(self, nums):
        res = [0] * len(nums)
        res[0], factor = 1, 1
        for i in range(1,len(nums)):
            res[i] = res[i-1] * nums[i-1]
        for i in range(len(nums)-1, -1, -1):
            res[i] *= factor
            factor *= nums[i]
        return res 

File: cpp/leetcode/238.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size(), fromBegin = 1, fromLast = 1;
        vector<int> res(n, 1);
        for(int i = 0; i < n; i++){
            res[i] *= fromBegin;
            fromBegin *= nums[i];
            res[n - i - 1] *= fromLast;
            fromLast *= nums[n - i - 1];
        }
        return res;
    }
};